src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs

```diff
@@ -0,0 +1,298 @@
+                            dte.StatusBar.Text = string.Format(ServicesVSResources.Installing_package_0_completed, packageName);
+                        }
+
+                        return true;
+                    }
+                    catch
+                    {
+                        dte.StatusBar.Text = string.Format(ServicesVSResources.Installing_package_0_failed, packageName);
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        internal void Connect(VisualStudioWorkspaceImpl workspace)
+        {
+            this.AssertIsForeground();
+
+            _workspace = workspace;
+
+            var componentModel = workspace.GetVsService<SComponentModel, IComponentModel>();
+            _packageInstallerServices = componentModel.GetService<IVsPackageInstallerServices>();
+            _packageInstaller = componentModel.GetService<IVsPackageInstaller>();
+
+            _workspace.WorkspaceChanged += OnWorkspaceChanged;
+        }
+
+        internal void Disconnect(VisualStudioWorkspaceImpl workspace)
+        {
+            this.AssertIsForeground();
+
+            Debug.Assert(workspace == _workspace);
+            _workspace.WorkspaceChanged -= OnWorkspaceChanged;
+            _workspace = null;
+        }
+
+        private void OnWorkspaceChanged(object sender, WorkspaceChangeEventArgs e)
+        {
+            // Can be called on any thread.
+            bool localSolutionChanged = false;
+            ProjectId localChangedProject = null;
+            switch (e.Kind)
+            {
+                default:
+                    // Nothing to do for any other events.
+                    return;
+                case WorkspaceChangeKind.ProjectAdded:
+                case WorkspaceChangeKind.ProjectChanged:
+                case WorkspaceChangeKind.ProjectReloaded:
+                case WorkspaceChangeKind.ProjectRemoved:
+                    localChangedProject = e.ProjectId;
+                    break;
+
+                case WorkspaceChangeKind.SolutionAdded:
+                case WorkspaceChangeKind.SolutionChanged:
+                case WorkspaceChangeKind.SolutionCleared:
+                case WorkspaceChangeKind.SolutionReloaded:
+                case WorkspaceChangeKind.SolutionRemoved:
+                    localSolutionChanged = true;
+                    break;
+            }
+
+            lock (_gate)
+            {
+                // Cancel any existing update task.
+                _tokenSource.Cancel();
+                _tokenSource = new CancellationTokenSource();
+
+                _solutionChanged |= localSolutionChanged;
+                if (localChangedProject != null)
+                {
+                    _changedProjects.Add(localChangedProject);
+                }
+
+                // Because we may get a lot of events, we attempt to throttle things so that
```

Comment:
This would be a bit clearer if the cancellation above and the enqueueing of the new task were together and you could say "Cancel the pending update task if it exists and enqueue a new update task that encompasses all of the pending changes." or something. Then, under the lock you have 1) Update pending changes lists, and 2) update task/wait state.